thermal.c

```diff
@@ -14,43 +14,333 @@
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: message setup failed.\n");
+		goto err_out;
+	}
+
+	cloned_callback = nl_cb_clone(callback);
+	if (!cloned_callback) {
+		log(TO_ALL, LOG_ERR, "thermal: callback handle duplication failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_send_auto_complete(sock, msg);
+	if (rc < 0) {
+		log(TO_ALL, LOG_ERR, "thermal: failed to send the first message.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_err(cloned_callback, NL_CB_CUSTOM, handle_error, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: error callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_ACK, NL_CB_CUSTOM, handle_end, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: ack callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_FINISH, NL_CB_CUSTOM, handle_end, &callback_rc);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: finish callback setup failed.\n");
+		goto err_out;
+	}
+
+	rc = nl_cb_set(cloned_callback, NL_CB_VALID, NL_CB_CUSTOM, handle_groupid, &nldata);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: group id callback setup failed.\n");
+		goto err_out;
+	}
+
+	while (callback_rc != 0) {
+		rc = nl_recvmsgs(sock, cloned_callback);
+		if (rc < 0) {
+			log(TO_ALL, LOG_ERR, "thermal: failed to receive messages.\n");
+			goto err_out;
+		}
+	}
 
-	log(TO_ALL, LOG_ERR, "thermal: not yet implemented to establish netlink.\n");
+	group_id = nldata.id;
+	if (group_id < 0) {
+		log(TO_ALL, LOG_ERR, "thermal: invalid group_id was received.\n");
+		goto err_out;
+	}
+
+	rc = nl_socket_add_membership(sock, group_id);
+	if (rc) {
+		log(TO_ALL, LOG_ERR, "thermal: failed to join the netlink group.\n");
+		goto err_out;
+	}
+
+	goto out;
+
+err_out:
+	error = TRUE;
+out:
+	nl_cb_put(cloned_callback);
+	nlmsg_free(msg);
 	return error;
 }
 
-static gboolean register_netlink_handler(nl_recvmsg_msg_cb_t handler __attribute__((unused)))
+static int handle_thermal_event(struct nl_msg *msg __attribute__((unused)),
+				void *arg __attribute__((unused)))
+{
+	log(TO_ALL, LOG_ERR, "thermal: not yet implemented to process thermal event.\n");
+	return NL_SKIP;
+}
+
+static int handler_for_debug(struct nl_msg *msg __attribute__((unused)),
+			     void *arg __attribute__((unused)))
+{
+	return NL_SKIP;
+}
+
+static gboolean register_netlink_handler(void)
 {
-	gboolean error = TRUE;
+	gboolean error = FALSE;
+	int rc;
+
+	rc = nl_cb_set(callback, NL_CB_SEQ_CHECK, NL_CB_CUSTOM, handler_for_debug, NULL);
+	if (rc) {
+		error = TRUE;
+		log(TO_ALL, LOG_ERR, "thermal: debug handler registration failed.\n");
+		goto out;
+	}
 
-	log(TO_ALL, LOG_ERR, "thermal: not yet implemented to register thermal handler.\n");
+	rc = nl_cb_set(callback, NL_CB_VALID, NL_CB_CUSTOM, handle_thermal_event, NULL);
+	if (rc) {
+		error = TRUE;
+		log(TO_ALL, LOG_ERR, "thermal: thermal handler registration failed.\n");
+	}
+out:
 	return error;
 }
 
+#define INVALID_NL_FD	-1
```

Comment:
what is this doing here?

Reply:
Given this, I thought the code is more readable than doing like this down there:

	fd = nl_socket_get_fd(sock);
	if (fd == -1) { ...

Reply:
I'm fine with that, its just in the middle of the code.  Please move it to the top of the file or in front of the function in which you use it

Reply:
Oh, I see. Will do that.